src/MadnetFactory.sol

```diff
@@ -0,0 +1,397 @@
+    */
+    function deployTemplate(bytes calldata _deployCode) public onlyOwner returns (address contractAddr) {
+        assembly{
+            //get the next free pointer
+            let basePtr := mload(0x40)
+            let ptr := basePtr
+            //codesize, PC,  pc, codecopy, codesize, push1 09, return push2
+            mstore(ptr, shl(168, or(0x38585839386009f3610000, _deployCode.length)))
+            ptr := add(ptr, 0x0b)
+            // modify runtime to contain the tail jump operation
+            // <codesize other> 56 5b 
+            // account for array offset
+            mstore8(ptr, 0x56)
+            ptr := add(ptr, 0x01)
+            mstore8(ptr, 0x5b)
+            ptr := add(ptr, 0x01)
+            //copies the initialization code of the implementation contract
+            calldatacopy(ptr, add(0x05, _deployCode.offset), sub(_deployCode.length, 0x05))
+            // Move the ptr to the end of the code in memory
+            // account for the previously added values to offset the copy
+            // account for the need to change the constructor dynamic byte array
+            // store the length of the initcode modifier
+            ptr := add(ptr, sub(_deployCode.length, 0x05))
+            ptr := sub(ptr, 0x20)
+            mstore(ptr, 0x28)
+            ptr := add(ptr, 0x20)
+            // finish the code with the terminate sequence  
+            // 5b 60 80 60 40 52 60 05 33 73 <factory> 14 15 57 33 ff fe
+            mstore8(ptr, 0x5b)
+            ptr := add(ptr, 0x01)
+            calldatacopy(ptr, _deployCode.offset, 0x05)
+            ptr := add(ptr, 0x05)
+            mstore(ptr, or(or(shl(240, 0x3373), shl(80, address())), shl(48, 0x14156004)))
+            ptr := add(ptr, 0x1a)
+            mstore(ptr, shl(192, 0x57361560045733ff))
+            ptr := add(ptr, 0x08)
+            contractAddr := create(0, basePtr, sub(ptr, basePtr))
+        }
+        codeSizeZeroRevert(uint160(contractAddr) != 0);
+        emit DeployedTemplate(contractAddr);
+        implementation_ = contractAddr;
+        return contractAddr;      
+    }
+
+    /**  
+    * @dev deployCreate deploys a contract from the factory address using create
+    * @param _deployCode bytecode to deploy using create
+    */
+    function deployCreate(bytes calldata _deployCode) public onlyOwner returns (address contractAddr) {
+        assembly{
+            //get the next free pointer
+            let basePtr := mload(0x40)
+            let ptr := basePtr
+            
+            //copies the initialization code of the implementation contract
+            calldatacopy(ptr, _deployCode.offset, _deployCode.length)
+
+            //Move the ptr to the end of the code in memory
+            ptr := add(ptr, _deployCode.length)
+            
+            contractAddr := create(0, basePtr, sub(ptr, basePtr))
```

Comment:
Are _deployCode.length and sub(ptr, basePtr)? If yes, maybe using _deployCode.length would be a little bit more clear.